### PR TITLE
[FEAT] Add SYNC button to withdraw modal

### LINKED ISSUE #573: [Proposal] Check sync-state prior to withdrawal function
There have been numerous reports in the Discord of individuals that are missing the text about sync needing to be done prior to withdrawal. This proposal, which I am hoping to use as my first contribution, is to add a sync-state check of some variety to the button action of the withdrawal button itself in the bank modal. Additionally, I would like to modify the text within that modal to be more "in your face" as a temporary measure during the design and testing of said function.

### PR DESCRIPTION
**Pull request by newb23**
# Description

Closes #573 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Testing was done locally, and re-based from the latest main just prior to PR.

Test Suites: 12 passed, 12 total
Tests:       103 passed, 103 total
Snapshots:   0 total
Time:        26.826 s
Ran all test suites.
Done in 27.93s.

image

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [x] Screenshot if it includes any UI changes
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by ahh1539** [comment]
Great change! Might I suggest that you add a little margin to the top of the sync button?

**Comment by adamhannigan** [comment]
@newb23 When a player sees this button they may think that they **need to** sync before withdrawing.

Do you think it would be possible to only show this if they have data that has not been synced?

**Comment by newb23** [comment]
> @newb23 When a player sees this button they may think that they **need to** sync before withdrawing.
> Do you think it would be possible to only show this if they have data that has not been synced?

@adamhannigan 

There are a couple of ways we could go about this. 1, I could add a localStorage option that will simply log when last the user `sync`d, or 2, that same data could be used plus a comparator, say, number of "saves" or "change events" since last sync, but all saved on the interim API server.

The first option is much like the one that @brian-man brought up in the issue I originally opened, but would lack the "transferability" of 2, wherein if they change between devices on option 1, that tracking is lost.

**Comment by adamhannigan** [comment]
@newb23 Is it possible that we do a dynamic check?

I.e. Load the blockchain data and compare it to our local balance + inventory. If they are different, it means they are out of sync with the blockchain.

We have some helper methods in the repo that load this data. We use a few of these in our gameMachine `loading` state that initialises the game state.

**Comment by newb23** [comment]
> @newb23 Is it possible that we do a dynamic check?
> 
> I.e. Load the blockchain data and compare it to our local balance + inventory. If they are different, it means they are out of sync with the blockchain.
> 
> We have some helper methods in the repo that load this data. We use a few of these in our gameMachine `loading` state that initialises the game state.

@adamhannigan 

Absolutely! A dynamic check is ideal, as one of the more important parts of option 2. I will check out the `gameMachine loading` state and these helper methods, then hide/disable/enable the buttons as needed depending on that state.

**Comment by newb23** [comment]
This is working locally, I am just struggling through the LINT process. :D More to come on it tomorrow. 

TL;DR - I have the local/saved data successfully checking against the chain sync'd data for both inventory and balance. This check has *only* been applied to the sync button within the Withdraw tab of the Bank, but, should be able to be ported to anywhere else desired pretty easily. Once I figure out Husky/LINT/Prettier issues...

**Comment by newb23** [comment]
Okay! Everything is now passing, and I have proof of LINT compliance and the actual `yarn test` passing text as well. I also redesigned the SYNC button a touch, so they are all part of one element, and the emtire element is hidden based on the comparative state of the local/saved state vs chain. AKA: You'll only ever see that warning and SYNC button if you're out of sync!

**Comment by adamhannigan** [comment]
@newb23 We have some new changes coming in with the Goblin town which restructures the syncing process (before they go to Goblin town). These changes will be going in shortly - it would be good to use this workflow before entering goblin town.

**Comment by ShiningMinions** [comment]
@newb23 Sorry, I know I'm late to the party here. First of all, I think this is an awesome change and adds a ton of value to the players. Second of all, I have one potential UX suggestion. (I know you've been patiently waiting for this PR to get merged, so maybe it can be in future PR or something.)

What do you think about also disabling the "withdraw" button when the player is out of sync? In other words, why even give the player an opportunity to screw up their game-state in the first place?

**Comment by newb23** [comment]
@ShiningMinions I am all for protecting from drunk-ups. My impression was that we didn't want to "force" the sync, but it is super simple to implement disabling the button, and/or just straight calling the sync method on use of the button, with a negative check on the isSynced function.

I likewise have no issue with PRs taking a while. It's not like they're only a matter of weeks out from initial release or anything. lol

**Comment by ShiningMinions** [comment]
@newb23 yeah it's up to you and what the others think on this PR as well. As a player myself, I honestly would prefer it to be this way so there's no chance I can even accidentally make a mistake here. Also look in #farmers-chat, there are so many people complaining every day about losing progress this way.

**Comment by newb23** [comment]
> @newb23 yeah it's up to you and what the others think on this PR as well. As a player myself, I honestly would prefer it to be this way so there's no chance I can even accidentally make a mistake here. Also look in #farmers-chat, there are so many people complaining every day about losing progress this way.

Trust me, I've been seeing it since day "1" (my day one, where I subsequently got caught in that whoopsy ban-wave, lol) where I was helping out in the various bug channels. It served a big reason for why I wanted to figure out this logic in the first place. I just want to be sure that: 
> When a player sees this button they may think that they need to sync before withdrawing. Do you think it would be possible to only show this if they have data that has not been synced?
Does not carry over to having it "literally" on the button, or if was just wanting proper sync-check first. I think the latter, but don't want to make improper assumptions.

Rambling, but, TL;DR, I concur and want, as a player, it just "just work", which in this case would be a sync (if needed) then withdraw all in one button.

**Comment by adamhannigan** [comment]
@newb23 This was an amazing initiative but we will need to plan out and rework this with the current Goblin town.

It would be awesome to have this check on the Goblin Entry Modal and only show the warnings if dynamically they have changes.

**Comment by newb23** [comment]
> @newb23 This was an amazing initiative but we will need to plan out and rework this with the current Goblin town.
> 
> It would be awesome to have this check on the Goblin Entry Modal and only show the warnings if dynamically they have changes.

@adamhannigan 
Hey sir. Now that I have had a good look on Goblin town, I can definitely rework it so that we can check pre-entrance. Additionally, I want to modularize said logic to include some of the other function areas, namely reset session and possibly some others I am not thinking of ATM, or for functions that will be added later.
